agent/src/agent/modules/functions/transform.py

```diff
@@ -0,0 +1,39 @@
+import re
+
+from typing import Callable
+
+TO_UPPER = 'to_upper'
+RE_SUB = 'regex_substring'
+
+# this is a number of arguments for each function that must be provided in the pipeline config
+# probably it's not the best place for them, think how to make validation better
+NUM_FUNCTION_ARGS = {TO_UPPER: 0, RE_SUB: 2}
+
+
+def to_upper(s: str) -> str:
+    return s.upper()
+
+
+def regex_substring(string: str, pattern: str, repl: str) -> str:
+    res = re.sub(pattern, repl, string)
+    if res == 'no_match':
+        raise Exception(f'Pattern `{pattern}` didn\'t match anything for the string `{string}`, repl `{repl}`')
+    return res
+
+
+def get_by_name(name: str) -> Callable:
+    if name == TO_UPPER:
+        return to_upper
+    elif name == RE_SUB:
+        return regex_substring
+    else:
+        raise Exception(f'Transformation function `{name}` is not supported')
+
+
+# todo this validation works during a pipeline run, not on creation
+def validate(func: Callable, args: list):
+    name = func.__name__
+    if name not in NUM_FUNCTION_ARGS:
+        raise Exception(f'Function `{name}` is not supported')
+    if NUM_FUNCTION_ARGS[name] != len(args):
```

Comment:
I think you can use inspect.signature module from a python standard library to get the number of arguments in a function [https://stackoverflow.com/questions/847936/how-can-i-find-the-number-of-arguments-of-a-python-function](url)

Reply:
The thing is, a user doesn't need to provide all the arguments. For example, regex_substrig takes 3 args, but the user needs to provide only 2 because the transformed value itself is the third argument. That's why I made it as a separate configuration

Reply:
Maybe when you call validate method you can pass all arguments? or maybe it's just always n_args-1

Reply:
yes, also thought about it, will use minus one